relation_engine_server/utils/ensure_specs.py

```diff
@@ -0,0 +1,265 @@
+"""
+Ensure that all the specs in the spec/**/*.json and spec/**/*.yaml are
+present in the server, with the top-level fields of the local specs being
+a subset of the top-level fields of the server specs
+"""
+import json
+from typing import Union, Callable
+
+from relation_engine_server.utils.json_validation import load_json_yaml
+from relation_engine_server.utils import arango_client
+from spec.validate import get_schema_type_paths
+
+
+def get_local_coll_indexes():
+    """
+    Read all schemas for the collection schema type
+    Return just collection name and indexes
+    """
+    coll_spec_paths = []
+    coll_name_2_indexes = {}
+    for coll_spec_path in get_schema_type_paths("collection"):
+        coll = load_json_yaml(coll_spec_path)
+        if "indexes" not in coll:
+            continue
+        coll_spec_paths.append(coll_spec_path)
+        coll_name_2_indexes[coll["name"]] = coll["indexes"]
+    return coll_spec_paths, coll_name_2_indexes
+
+
+def ensure_indexes():
+    """
+    Returns tuple
+    First item is list of borked index names, e.g.
+    [
+        "coll_name_3/fulltext/['scientific_name']",
+        "coll_name_4/persistent/['id', 'key']",
+    ]
+    Second item is struct of failed indexes, e.g.,
+    {
+        coll_name_3: [
+            {"type": "fulltext", "fields": ["scientific_name"] ...}
+        ],
+        coll_name_4: [
+            {"type": "persistent", "fields": ["id", "key"] ...}
+        ]
+    }
+    """
+    coll_name_2_indexes_server = arango_client.get_all_indexes()
+    coll_spec_paths, coll_name_2_indexes_local = get_local_coll_indexes()
+
+    failed_specs = {}
+    for coll_spec_path, (coll_name, indexes_local) in zip(
+        coll_spec_paths, coll_name_2_indexes_local.items()
+    ):
+        print(f"Ensuring indexes for {coll_spec_path}")
+        if coll_name not in coll_name_2_indexes_server:
+            failed_specs[coll_name] = indexes_local
+            continue
+        else:
+            failed_specs[coll_name] = []
+        indexes_server = coll_name_2_indexes_server[coll_name]
+        for index_local in indexes_local:
+            match = False
+            for index_server in indexes_server:
+                if index_local.items() <= index_server.items():
+                    match = True
+                    break
+            if match is False:
```

Comment:
this snippet is repeated in similar forms in the other `ensure_*` functions -- can you factor it out?